schema/query/predicate_validator.go

```diff
@@ -52,29 +63,46 @@ func validateValue(field string, value Value, validator schema.Validator) error
 		return err
 	}
 	if f.Validator != nil {
-		if f.Validator != nil {
-			if _, err := f.Validator.Validate(value); err != nil {
-				return fmt.Errorf("%s: invalid query expression: %s", field, err)
-			}
+		if _, err := f.Validator.Validate(value); err != nil {
+			return fmt.Errorf("%s: invalid query expression: %s", field, err)
 		}
 	}
 	return nil
 }
 
-func validateNumericValue(field string, value Value, op string, validator schema.Validator) error {
+func prepareValues(field string, values []Value, validator schema.Validator) ([]Value, error) {
 	f, err := getValidatorField(field, validator)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	if f.Validator != nil {
-		switch f.Validator.(type) {
-		case *schema.Integer, *schema.Float, schema.Integer, schema.Float:
-			if _, err := f.Validator.Validate(value); err != nil {
-				return fmt.Errorf("%s: invalid query expression: %v", field, err)
-			}
-		default:
-			return fmt.Errorf("%s: cannot apply %s operation on a non numerical field", field, op)
+		if preparator, ok := f.Validator.(schema.FieldPreparator); ok {
+			for i, value := range values {
+				nv, err := preparator.Prepare(value)
+				if err != nil {
+					return nil, err
 		}
+				values[i] = nv
```

Comment:
Same, can we reuse `values` here, and not defining `newValues`?

Reply:
no, need to define new values.